editor/src/messages/portfolio/document/node_graph/node_properties.rs

```diff
@@ -1835,6 +1837,77 @@ pub fn math_properties(node_id: NodeId, context: &mut NodePropertiesContext) ->
 	]
 }
 
+pub fn value_properties(node_id: NodeId, context: &mut NodePropertiesContext) -> Vec<LayoutGroup> {
+	let Some(document_node) = context.network_interface.document_node(&node_id, context.selection_network_path) else {
+		log::warn!("Value properties failed to be built because its document node is invalid.");
+		return vec![];
+	};
+	let Some(input) = document_node.inputs.get(0) else {
+		log::warn!("Secondary value input could not be found on value properties");
+		return vec![];
+	};
+	let mut select_value_widgets = Vec::new();
+
+	select_value_widgets.push(TextLabel::new("Value type: ").tooltip("Select the type the value node should output").widget_holder());
+
+	let Some(input_value) = input.as_non_exposed_value() else {
+		log::error!("Primary value node input should be a hidden value input");
+		return Vec::new();
+	};
+
+	let input_type = input_value.ty();
+
+	let Some(choice) = TaggedValueChoice::from_tagged_value(input_value) else {
+		log::error!("Tagged value in value node should always have a choice. input: {:?}", input);
+		return Vec::new();
+	};
+
+	// let committer = || {|_| {
+	// 	let messages = vec![
+	// 		DocumentMessage::AddTransaction.into(),
+	// 		NodeGraphMessage::RunDocumentGraph.into(),
+	// 	];
+	// 	Message::Batched(messages.into_boxed_slice()).into()
+	// };
+	let updater = || {
+		move |v: &TaggedValueChoice| {
+			let value = v.to_tagged_value();
+
+			let messages = vec![NodeGraphMessage::SetInputValue { node_id, input_index: 0, value }.into(), NodeGraphMessage::SendGraph.into()];
+
+			Message::Batched(messages.into_boxed_slice()).into()
+		}
+	};
+	let value_dropdown = enum_choice::<TaggedValueChoice>().dropdown_menu(choice, updater, || commit_value);
+	select_value_widgets.extend_from_slice(&[Separator::new(SeparatorType::Unrelated).widget_holder(), value_dropdown]);
+
+	let mut type_widgets = match property_from_type(node_id, 0, &input_type, (None, None, None), None, None, None, false, context) {
```

Comment:
I noticed that the following types don't have a working properties: OptionalF64, OptionalDVec2, VecU64, F64Array4.

Also the VectorData says "Vector data is supplied through the node graph". However this is misleading as there is no way to actually expose the input to the graph. This is also inconsistent with the other complex types e.g. GraphicElement, which currently just displays a "-".

Reply:
The `-` is newer and we can probably just standardize on that for all the types we don't yet have widgets for. We'll someday even have a widget for vector with some kind of way to visualize the path in a box that you can click on to view larger or something (I haven't thought much about this design, but that's the notion).